creator-node/src/snapbackSM/snapbackSM.js

```diff
@@ -377,37 +436,139 @@ class SnapbackSM {
+    let newReplicaNodesSet = new Set()
+    let selectNewReplicaSetAttemptCounter = 0
+    while (newReplicaNodesSet.size < unhealthyReplicasSet.size && selectNewReplicaSetAttemptCounter++ < MAX_SELECT_NEW_REPLICA_SET_ATTEMPTS) {
+      let randomHealthyNode = _.sample(healthyNodes)
 
-      // Select a random node that is not from the current replica set
-      newSecondary = healthyPotentialPeers[Math.floor(Math.random() * healthyPotentialPeers.length)]
+      // If node is already present in set, keep finding a unique healthy node
+      if (newReplicaNodesSet.has(randomHealthyNode)) continue
 
       // Check to make sure that the newly selected secondary does not have existing user state
       try {
-        const clockValue = await CreatorNode.getClockValue(newSecondary, wallet)
-        if (clockValue === -1) currentNodeHasExistingUserState = false
+        const clockValue = await CreatorNode.getClockValue(randomHealthyNode, wallet)
+        if (clockValue === -1) { newReplicaNodesSet.add(randomHealthyNode) }
       } catch (e) {
         // Something went wrong in checking clock value. Reselect another secondary.
-        this.logError(e.message)
+        this.logError(`[selectRandomReplicaSetNode] ${e.message}`)
       }
     }
 
-    if (currentNodeHasExistingUserState) throw new Error(`Unable to select new secondary given current healthy Content Nodes=[${healthyNodes}]`)
+    if (newReplicaNodesSet.size < unhealthyReplicasSet.size) {
+      throw new Error('[selectRandomReplicaSetNode] Not enough healthy nodes found to issue new replica set')
+    }
 
-    const currentHealthySecondary = unhealthyReplicasSet.has(secondary1) ? secondary2 : secondary1
+    return Array.from(newReplicaNodesSet)
+  }
 
-    // NOTE: For v0 implementation, `newSecondary2` will be the newly selected secondary, whilst the other nodes will be the
-    // original replica set primary and secondary1
-    return {
-      newPrimary: primary,
-      newSecondary1: currentHealthySecondary,
-      newSecondary2: newSecondary
+  // If the primary is part of the replica set, truly check to see if it is unhealthy.
+  async finalizeUnhealthyReplicaSet ({ unhealthyReplicasSet, primary, secondary1, secondary2, wallet, secondariesToClockMap }) {
+    let updatedUnhealthyReplicasSet = new Set(unhealthyReplicasSet)
+    // TODO: Should we query this multiple times??
+    if (updatedUnhealthyReplicasSet.has(primary)) {
+      try {
+        await this.peerSetManager.queryVerboseHealthCheck(primary)
+        updatedUnhealthyReplicasSet.delete(primary)
+      } catch (e) { /* Swallow error */ }
     }
+
+    // Fetch the clock values for the secondaries. This map is only used in the case that only the primary is unhealthy.
+    // However, if after `maxClockFetchAttempts` that a clock value cannot be retrieved, we should mark the secondary as unhealthy
+    // TODO: should we though? maybe the clock check could just be flakey.
```

Comment:
yeah i think we def should consider a secondary unhealthy if after multiple attempts the clock check is unresponsive

Reply:
yeah this is a good point we discussed before. I was also thinking along the lines of if someone updated the clock check and introduced some flakey code, the next two lines would introduce a false negative that the secondaries are unhealthy

Reply:
resolving bc outdated